### ISSUE TITLE
Exception using fn*, possibly related to use of closure?

### ISSUE THREAD

**Issue by sparkofreason** [atdixon/thurber#2]
Code:

```clojure
(defn pubsub-subscription->queue
  [subscription-name buffer-length]
  (let [q (ArrayBlockingQueue. buffer-length true)
        pubsub-source (.. (PubsubIO/readStrings)
                          (fromSubscription subscription-name))]
    (-> (th/create-pipeline {:streaming true :block-on-run false})
        (doto
         (th/apply!
          pubsub-source
          (th/fn* ->c [v] (.put q v) v))))))
```

Evaluating this gives 

"ck.beam_direct_runner_pubsub_fix$eval8276$pubsub_subscription__GT_queue__8277$eval8279__8280.<init>()"

Stack trace is a bunch of clojure eval stuff, first mention of thurber is thurber.clj line 133. Getting rid of the use of `q` fixes this. Workaround for now is to pull that code into a var and use `th/partial`.

**Comment by atdixon** [comment]
`th/fn*` cannot access lexical closure state. This is going to be a gotcha so I'll keep this issue open until I can get better error messaging in place and perhaps better documentation.

The reason is that Beam DoFns (and any referenced state) has to be Serializable so they can be distributed across the cluster. What `th/fn*` does is effectively define the function by name in the current namespace and references the var (as if you'd `(defn pubsub-source ...)` and referenced it as `#'pubsub-source`.)

There are a couple of ways around this:

* Declare the `q` state in the namespace (not as a local binding)
* (Only if `q` were serializable) you could relay the value via `th/partial`:

```
(th/partial (th/fn* pubsub-source [q v] ...) q)
```

As a repl demonstration of `th/fn*`:
```
(def xx 100)
(require 'thurber)
(thurber/fn* inlined-fn [] (+ 99 xx))
(#'inlined-fn)
=> 199
```

**Comment by sparkofreason** [comment]
Yeah, I figured closures might be an issue. Partial works.

**Comment by atdixon** [comment]
I've just added support (available as of release `0.0.4-alpha2`) for using _serializable_ values from lexically-scoped vars in `fn*`:

From the walkthrough:

```.clojure
(def simple-pipeline
  (let [multiplier 4] ;; multiplier can now be used in fn* (see below)
    (doto (th/create-pipeline)
      (th/apply! data-source
        (th/fn* multiply-fn [elem]
          (* elem multiplier))
        log-sink))))

;; This logs 4, 8, and 12:
(.run simple-pipeline)
```

When a `fn*` references vals in the lexical scope, a `th/partial` will be created under-the-covers. This means that combiners still won't accept these at the moment (per [issue/3](https://github.com/atdixon/thurber/issues/3)), but this will now work for ParDo usages.
